Marlin/src/HAL/STM32/HAL.h

```diff
@@ -138,7 +138,7 @@
 
 typedef double isr_float_t;   // FPU ops are used for single-precision, so use double for ISRs.
 
-#if defined(STM32G0B1xx) || defined(STM32H7xx)
+#if defined(STM32G0B1xx) || defined(STM32H7xx) || defined(STM32F401xC)
```

Comment:
Is this now an exhaustive list of all STM32 MCUs that require 32-bits for their pin numbers? I'm sure all processors have 32768 pins or fewer, but perhaps this 32-bits makes using the values faster due to ARM/RISC voodoo.

Reply:
that shouldnt be all STM32, only just those 3?

Reply:
`PinName` is just an untyped `enum` so I believe that makes its type the same size as `int`.

Where it all began…

The first STM32 HAL just used `uint8_t` for `pin_t`. Then it got bumped to 16 bits by #14523 (for STM32F407). Then in #23283 it got 32 bits for STM32G0B1xx, keeping 16 bits for others. Then #25387 added on STM32H7xx. With this PR adding STM32F401xC, it got me wondering.

STMicro's STM32 platform expects `uint32_t` for pin conversion functions like `digitalPinToPinName`. But `framework-arduino-hc32f46x` defines and uses `typedef int16_t gpio_pin_t` and `framework-arduinoststm32-maple` uses only 8 bits for functions like `gpio_set_pin`.

So for HC32 using Maple it may be more efficient to use 16 bit pin numbers. In general we want to find the best balance between size and speed. Different-sized types for pin numbers may or may not increase the size of the code and SRAM usage. But if a larger type results in faster (and potentially smaller) code then the tradeoff may be worthwhile.

Hmm…

Reply:
Interestingly, Marlin code size for a HAL/STM32 build is no different whether `pin_t` is defined with 8, 16, or 32 bits. So it seems best to just go with 32-bits for the best parity with `platforms/ststm32`.

Reply:
it would be a good idea for a 2nd opinion. @shadow578 seems to know HC32 in and out. also remember HC32 is Cortex-M4 and should be treated separately even though it takes cue from STM32F1-Maple.

the fix I found for the warning "overflow in conversion from 'PinName' to 'pin_t'" happened by using default_envs = `STM32F401RC_btt`. I haven't try seeing if pops up for any other STM32F401RC platforms, like Ender-3 S1.

so we could see through the Tests/Checks logs if that kind of warning shows up. I mean if that warning isn't present (before this **PR** change in question)

I guess what you're wondering is would it benefit ALL the chips? well, if it is good for one, it should be good for all?

then that may take some actual trial and error, or real world testing. if theres something I can do on my end just let me know.